### ISSUE TITLE
[Discussion] State of repository and some thoughts on changes

### ISSUE THREAD

**Issue by Ash258** [lukesampson/scoop#2609]
I would like to start discussion about few things I came up while using scoop as user and bucket maintainer. Sorted by complexity. Normally i would post it to project's Slack workspace / Discord, but gitter is not best option for this so I create issue.

- [x] `scoop.sublime-project`
    - All settings are covered in .editorconfig
    - Really someone use sublime in age of Visual Code / Atom? 🤔🤔

~~- [ ] Add scoops schema.json to [Schema store](https://github.com/SchemaStore/schemastore)~~
- [x] Hash variables
    - `core.ps1` could store some variables like `$hashMD5, $hashSHA1, ...` for substituion inside `find` property
        - Easier to write `<elementwithhash>($hashMD5)` than `<elementwithhash>([A-Fa-f\\d]{32})`
- [ ] Move All scoop related repositories into 1 organization (#3437)
    - ScoopInstaller/Extras, ScoopInstaller/Core, ScoopInstaller/Main, ScoopInstaller/Java, ScoopInstaller/Binaries, ...
    - No need to go through X github accounts to find buckets (java, versions, nerdfonts)
        - It does not look "official" and clean, when you are pulling / going through manifests from buckets, where you cannot tell if its "verified" or not.
        - Transfering all of them under Scoop organization will make it more reliable.
- [ ] Changelog property (#2067)
    - URL with variables substitution support.
- [ ] Some restrictions, rules to write manifests (#3362)
    - There are some manifests, which does not look good and are really messy.
    - Few fast searched examples:
        - [Jetbrains-toolbox](https://github.com/ScoopInstaller/Main/blob/master/bucket/jetbrains-toolbox.json)
            - It's post install code is mess.
            - Maybe historical reason??
        - [inadyn-mt](https://github.com/ScoopInstaller/Main/blob/master/bucket/inadyn-mt.json)
            - Complete mess, without some logical organization
            - Url as last element?
            - Hash element location
            - First autoupdate?
            - ...
        - [cdrtools](https://github.com/ScoopInstaller/Main/blob/master/bucket/cdrtools.json)
            - Again illogical organization
- [ ] Manifests variables
    - Would be nice to have new property (named `variables`).
        - It would be object, with name of specific variable and value will be string
        - So we could define "globals" variables and use them inside all script blocks
            ```json
            {
                "variables":{
                    "somefile" : "$some\\path\\to\\file.ps1",
                    "content": "Get-Content $otherFile" // Allow substitution
                },
                "pre_install":[
                    "Set-Content $someFile $someContent"
                ],
                "uninstaller":{
                    "script": [
                        "Remove-Item $someFile -Force"
                    ]
                }
            }
            ```

- [ ] Shimming
    - [ ] ps1 files
        - While shimming ps1 files, help comment at start of file (Synopsis, parameters, ..), param block should be in final shimmed ps1 file.
        - Making powershell intellisense work and doing it's job.
    - [ ] exe
         - Keep icon while shimmed exe contains icon
- [x] Create [generic bucket repository](https://github.com/Ash258/GenericBucket)
    - Clean repository without manifest only for forking.
        - Containing binaries, test files, ..
        - Universal and all projects wide configuration
            - .vscode folder with some snippets, ...
    - Easier for new users to create own buckets with full experience of validating via pipeline
    - Updating it's files accordingly
    - As a custom bucket owner who has some responsibility for my customers, I like to have everything under control (pipeline checking).
    - When you fork repository you have that link for upstream and could easily update these files.
- [x] Binary for hashes recalculation / extraction for all urls, not only first. (when using checkver -f) (#2348)
- [x] Extracting main bucket from scoop source (#1328)
    - It would need some components rewriting, but would be much cleaner.
    - Having main bucket inside `buckets` dir not in `app/scoop/current/bucket`
    - Also extract supporting\shimexe and validator projects into own projects
        - No need to have project files inside clients scoop directory
        - Single binary for shimexe and validator should be enough
- [ ] Are there some requirements for `buckets.json` file?
- [ ] Scoop commands
    - It would be useful if commands (scoop list, scoop info, ...) returns some data.
    - That write-hosts are nice for users, but not much for "developers", because it's really pain to capture write-hosts.
    - When you want to check (some script, manifests, checkver, ...) if app is installed you need to to something like `(gci "$env:SCOOP\apps" -Filter 'app').Exists`
        - Instead of some intuitive way `(scoop info <app>).Installed`, `(scoop list) -contains 'app'`, ...
- [x] Speaking of checkver (#2624)
    - Checkver could have switch for printing, returning only outdated manifests.
        - It's really messy when you have bigger bucket.
- [ ] Support / Move to yml.
    - ✋ Blockers:
        - [x] lukesampson/scoop#2939
        - [ ] lukesampson/scoop#3003
        - [x] cloudbase/powershell-yaml#38
    - Not sure how hard it would be.
        - It should not be that hard.
        - Only need to Install-Module https://github.com/cloudbase/powershell-yaml
        - Change `ConvertFrom-Json` to `ConvertFrom-Yaml`
    - Tons of advantages
        - Multiline scripts `key: |`
            - https://github.com/Ash258/scoop-Ash258/blob/2935029180d6e722ac47837ab80ba88da8c69361/appveyor.yml#L22-L29
        - COMMENTS SUPPORT ❗❗❗
        - No need to escape usual characters,
        - Better for maintaning
        - More readable
- [ ] Search
    - Definitely possible with https://github.com/pakeweb/scoop.sh
        - Just add new argument and simple WebClient to download response and print to console
    - There should be switch, which search all rapositories, not only locals.
        - Only dumb API caller.
        - It would require to have some sort of online API, which look for all git repositories with `scoop-bucket` tag and return all matched manifests, not depends on local changes.

Otherwise. Keep going. Really appreciate and enjoy scoop 😍

**Comment by Calinou** [comment]
> There are some manifests, which does not look good and are really messy.

We could use a tool to format and sort properties in all manifests at once.

**Comment by Ash258** [comment]
Sure. .\bin\format.ps1 doing its job with formating. But sorting / organizing properties would require to parse json files into some Class (in best way) with all it's properties and then just print it to file.

**Comment by kiedtl** [comment]
> That write-hosts are nice for users, but not much for "developers", because it's really pain to capture write-hosts.

I can't agree with that. Scoop should be built with ease of use for the end user, not a developer.

Yes, it's a pain to capture write-hosts, and prior to PSH 5, impossible. But there are plenty of [ways around this](https://github.com/lptstr/winfetch/blob/0fbaf75b3cd7e24b559a5df60582049b1f3fc011/src/winfetch.ps1#L241).

**Comment by r15ch13** [comment]
Renaming examples:
- [`scoopinstaller/scoop-main`](https://github.com/scoopinstaller/scoop-main) > `scoopinstaller/Main`
- [`scoopinstaller/php`](https://github.com/scoopinstaller/php) > `scoopinstaller/PHP`
- [`scoopinstaller/versions`](https://github.com/scoopinstaller/versions) > `scoopinstaller/Versions`
- [`scoopinstaller/binary-mirror`](https://github.com/scoopinstaller/binary-mirror) > `scoopinstaller/Binary`
- [`scoopinstaller/scoop-clone`](https://github.com/scoopinstaller/scoop-clone) > `scoopinstaller/Clone` (or other fancy name)
- [`scoopinstaller/shimexe`](https://github.com/lukesampson/scoop/tree/master/supporting/shimexe) > `scoopinstaller/Shim` (extraction required)
- [`scoopinstaller/validator`](https://github.com/lukesampson/scoop/tree/master/supporting/validator) > `scoopinstaller/Validator` (extraction required)
- [`scoopinstaller/install`](https://github.com/scoopinstaller/install) > `scoopinstaller/Install`
- [`scoopinstaller/nightlies`](https://github.com/scoopinstaller/nightlies) > `scoopinstaller/Nightlies` (delete because it's unused or reactivate)
- [`scoopinstaller/awesome-scoop`](https://github.com/scoopinstaller/awesome-scoop) > `scoopinstaller/Awesome`
- [`scoopinstaller/scoop.sh`](https://github.com/scoopinstaller/scoop.sh) > leave as is

**Comment by AndrewSav** [comment]
Not sure about yml. Having worked with it last two years, got a lot of problems from developers who find it difficult, to get the indentation right, get string escaping right and distinguish between lists and properties.
On the topic of strings there are [nine different ways](https://stackoverflow.com/a/21699210/284111) to write them with yaml, which, again, from experience, developers find confusing.

What are specific problems with json (apart from not being able to write comments?).

[Toml](https://github.com/toml-lang/toml) could be a simpler alternative to yaml, however it does not seem to have adequate tooling for Powershell.

**Comment by Calinou** [comment]
> What are specific problems with json (apart from not being able to write comments?).

JSON also doesn't support multi-line strings, so you have to emulate them with arrays.

**Comment by Ash258** [comment]
YAML is best suitable for manifest creation.

You do not quote any string in yaml. Unless it is number, which needs to be string

Here is comparison of manifest inside json and yaml
[json](https://github.com/Ash258/scoop-Ash258/blob/fec8b459a72c7dce91ecb78cedffb80ac753ba7c/bucket/FRD.json)
[yml](https://gist.github.com/Ash258/221b68710d5e0f7c667d9fdd07e1cda8)

Yml is just awesome for manifest

**Comment by h404bi** [comment]
> Toml could be a simpler alternative to yaml, however it does not seem to have adequate tooling for Powershell.

Acutally there are some toml libraries for .NET, such as [Nett](https://github.com/paiden/Nett). And I prefer toml too (see, Rust/Cargo is great!), yaml is simpler though.

**Comment by shoogle** [comment]
> YAML is best suitable for manifest creation.

You need to be careful with YAML. It certainly looks a lot nicer than JSON, but it's actually a hugely complicated format and a nightmare to parse, so most parsers don't implement the full feature set.

Consider the fact that YAML supports comments. At first glance this seems like an advantage, but let's see what would happen if you actually tried to use them in a manifest:

```yaml
description: Example YAML manifest
version: 2.3.1
# really important comment that maintainers absolutely must read!
checkver:
  url: https://example.com/download
```

Now let's run autoupdate:

```yaml
description: Example YAML manifest
version: 2.3.2
checkver:
  url: https://example.com/download
```

Oh dear, the comment has disappeared!

How can we avoid this happening? Like this:

```yaml
description: Example YAML manifest
version: 2.3.1
_comment: really important comment that maintainers absolutely must read!
checkver:
  url: https://example.com/download
```

This is exactly how comments in JSON work. The fact that JSON doesn't support proper comments *is a feature of the language*. The situation with newlines is similar, though not a serious.

The problem with comments is not unique to YAML, but it demonstrates that having another feature is not necessarily a good thing (and YAML has a *lot* of features!). If you want to switch to YAML then you should restrict yourself to using a sensible subset of its features.

**Comment by AndrewSav** [comment]
>Now let's run autoupdate:
What's autoupdate?

**Comment by kiedtl** [comment]
> Now let's run autoupdate:
> ```
> description: Example YAML manifest
> version: 2.3.2
> checkver:
> url: https://example.com/download
> ```
> Oh dear, the comment has disappeared!

It should be possible to add a script that will automatically add these comments back.

> What's autoupdate?

Checkver/autoupdate is a tool that bucket admins use to automatically update the manifests. E.g. if go-lang releases a new update, the autoupdate tool will automatically download the release and update the download URL, hash, and version of the manifest. In the process of doing so, some of the fields can be re-arranged and comments will disappear.

**Comment by shoogle** [comment]
@AndrewSav 
> What's autoupdate?

Autoupdate is a tool to ensure manifests always point to the latest version without requiring manual intervention by scoop maintainers. If you look through the bucket history you will see that the vast majority of commits are updating manifests, so autoupdate is central to how scoop operates. Comments, like [this one](https://github.com/ScoopInstaller/Main/blob/e79999782e82cf4ec168eeff37fad68f2851af37/bucket/git.json#L2) in git's manifest, need to be preserved.

> your comment does not see to apply to the scoop use case.

How can you claim that when you don't know what autoupdate is?

> I personally think that there are other facets of yaml which may make its use cumbersome

I totally agree, but as you said elsewhere, there are no plans to use those features. I was pointing out that even the features that have been explicitly cited as advantages, and given as reasons to switch, have unintended consequences.

@kiedtl 
> It should be possible to add a script that will automatically add these comments back.

You are proposing to create a custom YAML parser that doesn't obey the standard, for no better reason than syntax highlighting.

If comments are important then they should be treated as part of the data. This is the point that JSON's inventor made when he deliberately removed comments from the JSON spec.

Fundamentally, what is the difference between comments and app descriptions? One is aimed at developers, the other at users, yet both are equally important. We don't comment out `checkver` and `autoupdate` just because users aren't supposed to see them.

**Comment by shoogle** [comment]
Ultimately I'm in favour of a switch to YAML, just as long as it is used sensibly. Even if you end up treating it like JSON and doing this:

```yaml
description: Example YAML manifest
version: 2.3.1
_comment: really important comment that maintainers absolutely must read!
checkver:
  url: https://example.com/download
post_install:
  - $doThis
  - $doThat
  - $doSomethingElse
```

I think we can all agree that is a massive improvement over:

```json
{
    "description": "Example JSON manifest",
    "version": "2.3.1",
    "_comment": "really important comment that maintainers absolutely must read!",
    "checkver": {
        "url": "https://example.com/download"
    },
    "post_install": [
        "$doThis",
        "$doThat",
        "$doSomethingElse"
    ]
}
```

The problem with JSON is its unnecessary use of quotes and braces, not its less-is-more philosophy on features. It's just think it's a pity that the YAML developers couldn't content themselves with sorting out JSON's problems and instead went on to create a bunch of problems that didn't previously exist.

**Comment by AndrewSav** [comment]
>How can you claim that when you don't know what autoupdate is?

I'm right back with you. Now it makes total sense.

**Comment by mottosso** [comment]
As someone who recently discovered Scoop only a few days ago, I really can't comment much about which manifest format is best suited; however I wanted to see whether you've considered Powershell?

We're using a project similar to Scoop called [Rez](https://github.com/nerdvegas/rez) where YAML was the de-factor standard for the first few years, until it switched to Python (in which the project is also written). There was the argument, like with any configuration-like format, that using a scripting language gives the author too much power and invites complexity, but in reality what manifests do *is* complex. I think the issue of multiline is one example of this, and `post_install` which appear suitable for functions rather than plain-text like. Only now are we unconstrained by the format and can leverage the Python community to encapsulate just about any software project.

As an example.

- https://github.com/predat/rez-packages/blob/master/softs/maya/2018.0.4/package.py

On the flip-side, parsing a JSON is really straightforward and fast, and importantly doesn't carry side-effects. Something the manifest author would have to take into account if something like `scoop search` is going to be running `.ps1` scripts on the local (or remote!) machine just to get a hold of the metadata.

TLDR; a complex problem may rightfully require a complex solution

**Comment by shoogle** [comment]
The advantage of using a serialisation format rather than a scripting language is that it doesn't allow arbitrary code to be executed. I see this as a feature rather than a bug.

`post_install` is really an exception for applications that require extra configuration (e.g. to hook into a system service), but it is usually used (or abused) as a workaround for applications that have not been written properly (i.e. in a portable/generic way). I think `scoop install` should display a warning if a manifest contains a `post_install` step, and there should be an option to skip that step or even globally disable installation of applications that make use of it.

**Comment by kidonng** [comment]
Hi, since https://github.com/pakeweb/scoop.sh hasn't been active for a long time, I built another documentation site (inspired by the former one). Currently, it consists of two parts:

- Docs: basically a mirror of the latest wiki, with little tweaks to the contents.
- Apps: blazingly fast real-time search (thanks to Algolia) for [known buckets and some other buckets](https://github.com/kidonng/scoop-docs/blob/1c1c54941d487948e7c63a05b73f662de42f8289/scripts/index.js#L24-L32) 😃 search in https://github.com/rasa/scoop-directory really is inconvenient.

Welcome to share your advice on enhancing the docs site 😋

**Comment by dillonKneeland** [comment]
Hello, I've been using scoop for around a year now and something that I've always wished for was the ability to use `scoop search -Bucket <bucket-name>` to search for a program in a specific bucket or to be used without a query to list the apps in a bucket. It's something that I have working in my local install of scoop but is liable to break or get overwritten with every scoop update. I would be happy to make a PR for this feature if everyone agrees that it is as useful as I have found it to be.

**Comment by bb010g** [comment]
I really hope Scoop doesn't switch to YAML. JSON is much more sane to parse, and honestly to write. (YAML multiline strings even have their own reference website: <https://yaml-multiline.info/>.)

If you want to change off JSON to something more human-readable, I'd recommend [TOML](https://github.com/toml-lang/toml) or [eno](https://eno-lang.org/). TOML is probably the better option here, and it has some nice .NET parsing options, including [Tommy](https://github.com/dezhidki/Tommy) (MIT), a single-file parser that should be usable via [`Add-Type -TypeDefinition`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/add-type?view=powershell-5.1) if necessary, and [Tomlyn](https://github.com/xoofx/Tomlyn) (BSD-2-Clause), which supports modification of the syntax tree with perfect style preservation if necessary and has a compatible minimum .NET Framework requirement of 4.5 (for shipping the DLL).

<details><summary>Example <code>main\bucket\git.toml</code></summary>

```toml
version = "2.24.0.windows.2"
homepage = "https://gitforwindows.org"
description = "Distributed version control system."
license = "GPL-2.0-only"

[architecture.64bit]
url = "https://github.com/git-for-windows/git/releases/download/v2.24.0.windows.2/PortableGit-2.24.0.2-64-bit.7z.exe#/dl.7z"
hash = "353d0e1566d8897cb7afe2f6f9088bac17182ca43416feadec1c16f5c3bb9e0f"

[architecture.32bit]
url = "https://github.com/git-for-windows/git/releases/download/v2.24.0.windows.2/PortableGit-2.24.0.2-32-bit.7z.exe#/dl.7z"
hash = "6ac4761e2711cb584f2df1fbc27b677a03c745cfef0bae55604bc27a5c26d68b"

```

<details><summary><code># Comment</code></summary>

Here's `.bin` as it is now, converted to TOML:
```toml
bin = [
  'cmd\git.exe',
  'cmd\gitk.exe',
  'cmd\git-gui.exe',
  'usr\bin\tig.exe',
  'git-bash.exe',
]
```
Note that this would need to be declared at the root, above `[architecture]`.
(Tables cannot be defined more than once, and the root table doesn't even have
a name. It's more of an ambient key/value pair holder than a table.)

Some potentially more natural TOML encodings of `.bin` follow.

Encode `.bin` as a table (JSON object):

</details>

```toml
[bin.'cmd\git.exe']
[bin.'cmd\gitk.exe']
[bin.'cmd\git-gui.exe']
[bin.'usr\bin\tig.exe']
[bin.'git-bash.exe']

```

<details><summary><code># Comment</code></summary>

This is equivalent to the following inline table:
```toml
bin = {
  'cmd\git.exe' = {},
  'cmd\gitk.exe' = {},
  'cmd\git-gui.exe' = {},
  'usr\bin\tig.exe' = {},
  'git-bash.exe' = {},
}
```
Which is similar to the current array-based method, but you gain the benefit
of properties, as shown in `.shortcuts`.

Encode `.bin` as an array of tables:
```toml
[[bin]]
path = 'cmd\git.exe'
[[bin]]
path = 'cmd\gitk.exe'
[[bin]]
path = 'cmd\git-gui.exe'
[[bin]]
path = 'usr\bin\tig.exe'
[[bin]]
path = 'git-bash.exe'
```

I think this still works better than the current weakly typed
"`stringOrArrayOfStringsOrAnArrayOfArrayOfStrings`" approach, but you have
to specify a path every time, so unless we need to make duplicate shims
available, it makes sense to use tables keying on the path.

*With that said*, we still don't have to lose that advantage with the
table of tables approach. Allow `.bin[$path]` to be an array of tables as
well as a table, and you'd get this for a potential multiply-shimmed
program "bar" in a package providing "foo", "bar", and "baz" shims:
```toml
[bin.'foo.exe']
[[bin.'bar.exe']]
[[bin.'bar.exe']]
alias = 'bar3.exe'
[[bin.'bar.exe']]
alias = 'bar2.exe'
arguments = ['--two-mode']
[bin.'baz.exe']
```
Which looks pretty good, I'd say.

(This is also a perfect time to convert to arguments as an array instead of
a weirdly-escaped string. (lukesampson/scoop#2025))

We could put the `path` or `name` here in the key for `.shortcuts`, but I
used path for consistency with `.bin`.

</details>

```toml
[shortcuts.'git-bash.exe']
name = "Git Bash"
arguments = ['--cd-to-home']
[shortcuts.'cmd\git-gui.exe']
name = "Git GUI"
[shortcuts.'cmd\gitk.exe']
name = "gitk"

```

<details><summary><code># Comment</code></summary>

We can drop the array of strings dance here (and in `.pre_install`, and in
`.install.script`), because TOML has multi-line strings. (And they're good.)
Here we're using a multi-line literal string.

</details>

```toml
post_install = '''
git config --global credential.helper manager'''

[checkver]
github = "https://github.com/git-for-windows/git"
regex = '''v([\w.]+)/PortableGit-(?<ver>[\d\w.]+)-64-bit'''

[env_set]
GIT_INSTALL_ROOT = "$dir"

[autoupdate]
# note = """
# I'm a sample autoupdate note. \
# This is still the first line.
# Here's my second line!"""

[autoupdate.architecture.64bit]
url = "https://github.com/git-for-windows/git/releases/download/v$version/PortableGit-$matchVer-64-bit.7z.exe#/dl.7z"

[autoupdate.architecture.32bit]
url = "https://github.com/git-for-windows/git/releases/download/v$version/PortableGit-$matchVer-32-bit.7z.exe#/dl.7z"

[autoupdate.hash]
url = "https://github.com/git-for-windows/git/releases/latest"
regex = "<td>$basename</td>\\s*<td>$sha256</td>"
```

</details>

An advantage of using a library like Tomlyn for Scoop manifests would be that machine formatting could be specified, beyond just order of declarations, down to the canonical table syntax or string type for each property. You could make things *really* consistent.

(CC: Ash258/scoop#2)

**Comment by shoogle** [comment]
Or there's [StrictYAML](https://hitchdev.com/strictyaml/). It keeps the readability of YAML but removes the [unsafe and unnecessary features](https://hitchdev.com/strictyaml/).

**Comment by r15ch13** [comment]
@dillonKneeland I was working on https://github.com/lukesampson/scoop/pull/3446 and this could be added. But I haven't had the time to finish it yet.

@bb010g TOML looks really promising. 👍

**Comment by dwettstein** [comment]
Hey @Ash258

As you have closed this issue/discussion, what is the latest status of the Scoop project? It looks like it is getting less and less maintained. Additionally, I saw that you put a lot of effort into your fork (https://github.com/Ash258/Scoop-Core).

Furthermore, why are the remaining parts of Scoop not moved to the https://github.com/ScoopInstaller organization?

BR

**Comment by jtagcat** [comment]
I'd think this is the best place to mention.

Persons are interested in business support. I've seen multiple companies/people pass on scoop due to there being no support available. From the point of 'if something is broken, call/chat/mail support to have it fixed', and not use up a workday fixing things.

**Comment by domicioam** [comment]
Late to the party. I can see the project is being abandoned. Thank you @dwettstein for pointing that out.

**Comment by sharpninja** [comment]
I would like to add support for placing output in a PSObject instead of writing to stdout.  Would such a PR be welcome?

**Comment by rashil2000** [comment]
> I would like to add support for placing output in a PSObject instead of writing to stdout. Would such a PR be welcome?

@sharpninja Yes! Please find the relevant issue (or create one if it doesn't exist), and raise a PR against it :)

**Comment by sharpninja** [comment]
> > I would like to add support for placing output in a PSObject instead of writing to stdout. Would such a PR be welcome?
> 
> @sharpninja Yes! Please find the relevant issue (or create one if it doesn't exist), and raise a PR against it :)

I just finished up a bunch of changes for my PS Snippets and am motivated to work on this again.  I had these changes made at one point locally, but I'm not sure if they got lost with my laptop that died.  Will look for them.
